In [98]:
#Please install the tidyverse library by running this cell!
library(tidyverse)

# Reading in/importing Data Files
players_csv <- read_csv('players.csv')
sessions_csv <- read_csv('sessions.csv')
#players_csv
#sessions_csv

Rows: 196 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): experience, hashedEmail, name, gender
dbl (2): played_hours, Age
lgl (1): subscribe

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 1535 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): hashedEmail, start_time, end_time
dbl (2): original_start_time, original_end_time

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


### 1) **Data Desciption**
___

#### **Collection Method:**

Data was collected from a study done by the Pacific Laboratory for Artifical Intelligence (PLAI), a computer science research group led by Dr. Frank Wood. Their current project involves the creation an embodied AI model, which is an AI able to recieve, understand, and respond to a complex environment, similar to how humans interact with the world around them.

To do so, the popular game: "Minecraft" was chosen as the "complex environment". Behavioural data, such as locations visited, what players said in chat, in game actions, etc. were taken from players of the "PLAICRAFT" server, a free Minecraft server ran by PLAI, in order to study the trends and patterns of how players interact with the virtual world. This data will then be used to train and develop the AI model.

Information from the players.csv dataset was retreived primarily from the signup process of each player as well as some tracking of their play. On the contrary, all of the data besides the hashed E-mail used to identify the player, was retreived purely from player gameplay within the sessions.csv dataset. 

(references)

Pacific Laboratory for artificial intelligence. Pacific Laboratory for Artificial Intelligence. (2025, September 11). https://plai.cs.ubc.ca/

Pacific Laboratory of Artificial Intelligence. (2025). Plaicraft. PLAICraft. https://plaicraft.ai/

The University of British Columbia. Dr. Frank Wood. (n.d.). https://www.cs.ubc.ca/~fwood 

#### **Resultant Data:** <br>

#### For players.csv:

* **Description**

A dataset containing the information of players. Data includes the experience of the player, their subscription status, E-mail (note that it is hashed for security/privacy of participants), the total number of hours they played, name, gender, and age. Description of each variable is listed in the following table.

|Variable:|Data type:|Possible Values:|Desciption:|
|:----|:----|:----|:----|
| experience | character (chr) | Beginner, Amateur, Regular, Veteran, Pro | The self-declared skill level of the player |
| suscribe | logical (lgl) | TRUE, FALSE | Whether or not the player is suscribed |
| hashedEmail | character (chr) | N/A | A unique sequence of numbers and letters providing a secure way to represent a players E-mail |
| played_hours | double (dbl) | N/A | The number of hours a player has put into the server |
| name | character (chr) | N/A | The name of the player |
| gender | character (chr) | Male, Female, Non-binary, Two-Spirited, Prefer not to say, Other, Agender | Gender of the player |
| Age | double (dbl) | N/A | Age of the player |

<br>

* **Number of observations?**
  
    196, determined as there are 196 rows in the dataset

* **Number of Variables?**

    7, determined as there are 7 columns

* **Summary Statistics:** (Code/calculations in the following cell)

<br>

Distribution of experience levels:

|Value:|Percentage:|Count:|
|:----|:----|:----|
| Beginner | 17.86% | (35 players) |
| Amateur | 32.14% | (63 players) |
| Regular | 18.37% | (36 players) |
| Veteran | 24.49% | (48 Players) |
| Pro | 7.14% | (14 players) |

<br>Distribution of suscribed and unsuscribed players:

|Status:|Percentage:|Count:|
|:----|:----|:----|
| Suscribed | 73.47% | (144 players) |
| Unsuscribed | 26.53% | (52 players) |

<br>Lowest, Median, Mean, and Highest Playtime (hours):

|Lowest:|Median:|Mean:|Highest:|
|:----|:----|:----|:----|
| 0 | 0.1 | 5.85 | 233.1 |

<br>Distribution of Gender

|Gender:|Percentage:|Count:|
|:----|:----|:----|
| Male | 63.27 | (124 players) |
| Female | 18.88% | (37 players) |
| Non-Binary | 7.65% | (15 players) |
| Two-Spirited | 3.06% | (6 players) |
| Other | 0.51% | (1 players) |
| Agender | 1.02% | (2 players) |
| Prefer not to say | 5.61% | (11 players) |

<br>Lowest, Median, Mean, and Highest age (years):

|Lowest:|Median:|Mean:|Highest:|
|:----|:----|:----|:----|
| 9 | 19 | 21.14 | 58 |

<br>

* **Potential Issues:** <br>

Within the players.csv dataset there are very few potential issues. The main ones being `NA` values in the `Age` column as well as the `gender` and `experience` values being characters. The latter may need to be changed to a logical value to make computing or counting easier as to avoid having to filter for an exact string which my cause inefficiency within the analysis.

<br>

#### For sessions.csv:

* **Description**

A dataset containing the information of the player's gameplay sessions. Data includes the hashed E-mail (used to identify the player) as well as the start and end times of the players gameplay sessions in 2 formats. Description of each variable is listed in the following table.

|Variable:|Data type:|Possible Values:|Desciption:|
|:----|:----|:----|:----|
| hashedEmail | character (chr) | N/A | A unique sequence of numbers and letters providing a secure way to represent a players E-mail |
| start_time | character (chr) | N/A | The date and time of when a player began their play session |
| end_time | character (chr) | N/A | The date and time of when a player ended their play session |
| original_start_time | double (dbl) | N/A | The date and time of when a player began their play session in a double datatype format |
| original_end_time | double (dbl) | N/A | The date and time of when a player began their play session in a double datatype format |

<br>

* **Number of observations?**
  
    1535, determined as there are 1535 rows in the dataset

* **Number of Variables?**

    5, determined as there are 5 columns

* **Potential Issues:** <br>

Within the sessions.csv dataset there are a few issues which are needed to be taken into account. The first one is that the start_time and end_time contain both the time of the sessions as well as the date of the session, this is untidy and will need to be split. Secondly, the datatype of the `start_time` and `end_time` variables is character (chr) which will be needed to be converted to double (dbl) or integer (int) in order to perform calculations (e.g average play time in hours on x day). Lastly, the `original_start_time` and `original_end_time` are in a very starnge format which is hard to understand, conversion into a more drect format would benefit clarity of the data.

In [99]:
#Summary Statistics:

#--------------- Distribution of experience levels: ---------------#

#number of Beginner
num_Beginner <- players_csv |>
    filter(experience == 'Beginner') |>
    select(experience) |>
    nrow()
num_Beginner_perc <- round(((num_Beginner/196) * 100),2)
#num_Beginner
#num_Beginner_perc
    
#number of Amateur
num_Amateur <- players_csv |>
    filter(experience == 'Amateur') |>
    select(experience) |>
    nrow()
num_Amateur_perc <- round(((num_Amateur/196) * 100),2)
#num_Amateur
#num_Amateur_perc

#number of Regular
num_Regular <- players_csv |>
    filter(experience == 'Regular') |>
    select(experience) |>
    nrow()
num_Regular_perc <- round(((num_Regular/196) * 100),2)
#num_Regular
#num_Regular_perc

#number of Veteran
num_Veteran <- players_csv |>
    filter(experience == 'Veteran') |>
    select(experience) |>
    nrow()
num_Veteran_perc <- round(((num_Veteran/196) * 100),2)
#num_Veteran
#num_Veteran_perc

#number of Pro
num_Pro <- players_csv |>
    filter(experience == 'Pro') |>
    select(experience) |>
    nrow()
num_Pro_perc <- round(((num_Pro/196) * 100),2)
#num_Pro
#num_Pro_perc

#------- Distribution of suscribed and unsuscribed players: -------#

#number of suscribed
num_sub <- players_csv |>
    filter(subscribe == TRUE) |>
    select(subscribe) |>
    nrow()
num_sub_perc <- round(((num_sub/196) * 100),2)
#num_sub
#num_sub_perc

#number unsuscribed
num_unsub <- players_csv |>
    filter(subscribe == FALSE) |>
    select(subscribe) |>
    nrow()
num_unsub_perc <- round(((num_unsub/196) * 100),2)
#num_unsub
#num_sub_perc

#----------- Lowest, Median, Mean, and Highest Playtime: ----------#

#lowest playtime
low_pt <- players_csv |>
    select(played_hours) |>
    map_df(min, na.rm = TRUE) |>
    pull()
#low_pt

#median playtime
med_pt <- players_csv |>
select(played_hours) |>
    map_df(median, na.rm = TRUE) |>
    pull() |>
    round(2)
#med_pt

#mean playtime
mean_pt <- players_csv |>
    select(played_hours) |>
    map_df(mean, na.rm = TRUE) |>
    pull() |>
    round(2)
#mean_pt

#highest playtime
high_pt <- players_csv |>
    select(played_hours) |>
    map_df(max, na.rm = TRUE) |>
    pull()
#high_pt

#-------------------- Distribution of Gender: ---------------------#

#number of male
num_male <- players_csv |>
    filter(gender == 'Male') |>
    select(gender) |>
    nrow()
num_male_perc <- round(((num_male/196) * 100),2)
#num_male
#num_male_perc

#number of female
num_female <- players_csv |>
    filter(gender == 'Female') |>
    select(gender) |>
    nrow()
num_female_perc <- round(((num_female/196) * 100),2)
#num_female
#num_female_perc

#number of Non-binary
num_nb <- players_csv |>
    filter(gender == 'Non-binary') |>
    select(gender) |>
    nrow()
num_nb_perc <- round(((num_nb/196) * 100),2)
#num_nb
#num_nb_perc

#number of Two-Spirited
num_ts <- players_csv |>
    filter(gender == 'Two-Spirited') |>
    select(gender) |>
    nrow()
num_ts_perc <- round(((num_ts/196) * 100),2)
#num_ts
#num_ts_perc

#number of Other
num_oth <- players_csv |>
    filter(gender == 'Other') |>
    select(gender) |>
    nrow()
num_oth_perc <- round(((num_oth/196) * 100),2)
#num_oth
#num_oth_perc

#number of Agender
num_agn <- players_csv |>
    filter(gender == 'Agender') |>
    select(gender) |>
    nrow()
num_agn_perc <- round(((num_agn/196) * 100),2)
#num_agn
#num_agn_perc

#number prefer not to say
num_pns <- players_csv |>
    filter(gender == 'Prefer not to say') |>
    select(gender) |>
    nrow()
num_pns_perc <- round(((num_pns/196) * 100),2)
#num_pns
#num_pns_perc

#------------- Lowest, Median, Mean, and Highest Age: -------------#

#lowest age
low_age <- players_csv |>
    select(Age) |>
    map_df(min, na.rm = TRUE) |>
    pull()
#low_age

#median age
med_age <- players_csv |>
    select(Age) |>
    map_df(median, na.rm = TRUE) |>
    pull() |>
    round(2)
#med_age

#mean age
mean_age <- players_csv |>
    select(Age) |>
    map_df(mean, na.rm = TRUE) |>
    pull() |>
    round(2)
#mean_age

#highest age
high_age <- players_csv |>
    select(Age) |>
    map_df(max, na.rm = TRUE) |>
    pull()
#high_age

### 2) **Question**

Can the number of hours a player plays and their age predict their skill level?

Currently, from the players.csv dataset we have the variable of `played_hours`, `Age` and `experience` thus by using a knn model, we can predict the experience level of the player using the number of hours they played and their age as predictors. In terms of wrangling, there are a lot of players who signed up who didnt play, thus setting their hours played to 0 which need to be removed along with NA values in the `Age` variable in order to not skew the data.

### 3) **Exploratory Data Analysis and Visualisation**

The following cell contains:

1. Both datasets being loaded into R
2. Tidy versions of the sessions.csv (players.csv is already tidy)
3. The mean value of all quantitative variables in players.csv
4. Visualisations of each variable
5. Visualisation of played_hours

Some of these were previously calculated in part 1) but I added them here as well for clarity :)

Before reading in be sure to run the very first cell as it contains the necessary library (tidyverse)

In [100]:
#--------------------- 1) Loading in datasets: --------------------#

#loading in players.csv
players_csv <- read_csv('players.csv')
#players_csv

#loading in sessions.csv
sessions_csv <- read_csv('sessions.csv')
#sessions_csv

#--------------------- 2) Tidying sessions.csv: -------------------#


Rows: 196 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): experience, hashedEmail, name, gender
dbl (2): played_hours, Age
lgl (1): subscribe

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 1535 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): hashedEmail, start_time, end_time
dbl (2): original_start_time, original_end_time

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
